### PR TITLE
feat: allow configuring `use_legacy_protocol` per connection

### PR DESCRIPTION
**Pull request by thalesmg**
Related: https://github.com/emqx/emqx/discussions/11428

### CODE REVIEW

**Comment by zmstone** [comment]
https://github.com/emqx/mongodb-erlang/blob/2202840dad1c0ad0baef601521be80fc701d2627/src/main/mc_worker_pid_info.erl#L141-L146
maybe the check `[{_, #reply{documents = [#{<<"code">> := 352, <<"ok">> := 0.0}|_]}}|_]` is too assertive.
would be interesting what the reply is in mongodb v6.
or to be more future-proof, inverse the check to use `op_msg` for probing and then fallback to legacy if failed.

**Comment by thalesmg** [comment]
> https://github.com/emqx/mongodb-erlang/blob/2202840dad1c0ad0baef601521be80fc701d2627/src/main/mc_worker_pid_info.erl#L141-L146
> 
> maybe the check `[{_, #reply{documents = [#{<<"code">> := 352, <<"ok">> := 0.0}|_]}}|_]` is too assertive.
> would be interesting to find out what the reply is in mongodb v6.
> or to be more future-proof, inverse the check to use `op_msg` for probing and then fallback to legacy if failed.

Indeed, that would be an improvement.

If we give users a knob to tune in case things go wrong, they may resolve the issues themselves, though, so I think we still need this option.

We already test mongo 6.0 in CI, so I wonder if it's something in Mongo Atlas or a newer version :thinking:

**Comment by thalesmg** [comment]
Inverting the check/fallback is a good idea, I'm trying to make that change.

File: src/main/mc_worker_pid_info.erl [hunk 1]
```diff
@@ -117,32 +118,36 @@ install_mc_worker_info(Socket, NetModule, Database) ->
...
-            Response = mc_connection_man:request_raw_no_parse(Socket, Database, Request, NetModule),
-            case Response of
-                [{_, #reply{documents = [#{<<"code">> := 352, <<"ok">> := 0.0}|_]}}|_] ->
+            %% * https://www.mongodb.com/docs/manual/reference/command/hello/#mongodb-dbcommand-dbcmd.hello
+            %% * https://www.mongodb.com/docs/manual/reference/mongodb-wire-protocol/#std-label-wire-op-msg
+            {ok, Socket} = mc_worker_logic:connect_to_database(ConnectionOpts),
+            Command = bson:fields({hello, 1}),
+            Request = #op_msg_command{command_doc = Command, database = Database},
+            try mc_connection_man:request_raw_no_parse(Socket, Database, Request, NetModule) of
+                [{_, #op_msg_response{response_doc = #{<<"ok">> := 1.0}}} | _] ->
```
Comment (zmstone) [inline src/main/mc_worker_pid_info.erl:143]:
```suggestion
                [{_, #op_msg_response{response_doc = #{<<"ok">> := _}}} | _] ->
```

File: src/main/mc_worker_pid_info.erl [hunk 2]
```diff
@@ -117,32 +118,36 @@ install_mc_worker_info(Socket, NetModule, Database) ->
...
                     op_msg;
                 _ErrorResponse ->
                     legacy
+            catch
+                _:_ ->
```
Comment (zmstone) [inline src/main/mc_worker_pid_info.erl:148]: This won’t catch the try of clause mismatch.
Change to 
```
try [{_, #op_msg_response{response_doc = #{<<"ok">> := _}}} | _] = mc_connection_man:request_raw_no_parse(Socket, Database, Request, NetModule)
```

Comment (thalesmg) [reply src/main/mc_worker_pid_info.erl:148]: Sorry, which clause mismatch?  There's already a catch-all clause [here](https://github.com/emqx/mongodb-erlang/pull/39/files#diff-a1dae03dcf099259fce4c2ae4bc5c9555641c05f953a38835e35cde20e6a44a4R145).

**Review by zmstone — APPROVED** [review]

**Pull request merged**